cmd/qgb/relayer/cmd.go

```diff
@@ -1,20 +1,176 @@
+
+			// connecting to a QGB contract
+			ethClient, err := ethclient.Dial(config.evmRPC)
+			if err != nil {
+				return err
+			}
+			qgbWrapper, err := wrapper.NewQuantumGravityBridge(config.contractAddr, ethClient)
+			if err != nil {
+				return err
+			}
+
+			// creating Celestia-app configuration
+			encCfg := encoding.MakeConfig(app.ModuleEncodingRegisters...)
+
+			// creating an RPC connection to tendermint
+			trpc, err := http.New(config.tendermintRPC, "/websocket")
+			if err != nil {
+				return err
+			}
+			err = trpc.Start()
+			if err != nil {
+				return err
+			}
+			defer func(trpc *http.HTTP) {
+				err := trpc.Stop()
+				if err != nil {
+					logger.Error(err.Error())
+				}
+			}(trpc)
+
+			// creating tendermint querier
+			tmQuerier := rpc.NewTmQuerier(trpc, logger)
+			if err != nil {
+				return err
+			}
+
+			// creating a grpc connection to Celestia-app
+			qgbGRPC, err := grpc.Dial(config.celesGRPC, grpc.WithTransportCredentials(insecure.NewCredentials()))
+			if err != nil {
+				return err
+			}
+			defer func(qgbGRPC *grpc.ClientConn) {
+				err := qgbGRPC.Close()
+				if err != nil {
+					logger.Error(err.Error())
+				}
+			}(qgbGRPC)
+
+			// creating the application querier
+			appQuerier := rpc.NewAppQuerier(logger, qgbGRPC, encCfg)
+
+			// creating the host
+			h, err := libp2p.New()
+			if err != nil {
+				return err
+			}
+			logger.Info(
+				"created host",
+				"ID",
+				h.ID().String(),
+				"Addresses",
+				h.Addrs(),
+			)
+			// creating the data store
+			dataStore := dssync.MutexWrap(ds.NewMapDatastore())
+
+			// get the bootstrappers
+			var bootstrappers []peer.AddrInfo
+			if config.bootstrappers == "" {
+				bootstrappers = nil
+			} else {
+				bs := strings.Split(config.bootstrappers, ",")
+				bootstrappers, err = helpers.ParseAddrInfos(logger, bs)
+				if err != nil {
+					return err
+				}
+			}
+
+			// creating the dht
+			dht, err := p2p.NewQgbDHT(cmd.Context(), h, dataStore, bootstrappers, logger)
+			if err != nil {
+				return err
+			}
+
+			// wait for the dht to have some peers
+			err = dht.WaitForPeers(cmd.Context(), time.Hour, 10*time.Second, 1)
 			if err != nil {
 				return err
 			}
 
-			// TODO add implementation
+			// creating the p2p querier
+			p2pQuerier := p2p.NewQuerier(dht, logger)
+
+			relay, err := relayer.NewRelayer(
+				tmQuerier,
+				appQuerier,
+				p2pQuerier,
+				evm.NewClient(
+					logger,
+					qgbWrapper,
+					config.privateKey,
+					config.evmRPC,
+					config.evmGasLimit,
+				),
+				logger,
+			)
+			if err != nil {
+				return err
+			}
+
+			wg := &sync.WaitGroup{}
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for {
+					select {
+					case <-cmd.Context().Done():
+						return
+					default:
+						err = relay.Start(cmd.Context())
+						if err != nil {
+							logger.Error(err.Error())
+							time.Sleep(time.Second * 30)
+							continue
+						}
+						return
+					}
+				}
+			}()
+			wg.Wait()
```

Comment:
[blocking question]
what's the purpose of this loop? just to restart the relayer? is that a desired property? why is a goroutine needed here?

Reply:
definitly :+1: This is how it was in the previous implementation and didn't change it. Will make the change right away. Thanks

Reply:
[e862108](https://github.com/celestiaorg/orchestrator-relayer/pull/151/commits/e8621084c91d1e7376a95aef4c003bf9596fb4a6)

@evan-forbes Like this?